pkg/kubelet/api/testing/fake_runtime_service.go

```diff
@@ -19,66 +19,356 @@ package testing
+	sync.Mutex
+
+	Called []string
+
+	Containers map[string]*FakeContainer
+	Sandboxes  map[string]*FakePodSandbox
 }
 
-func (r *fakeRuntimeService) CreatePodSandbox(config *runtimeApi.PodSandboxConfig) (string, error) {
-	return "", fmt.Errorf("not implemented")
+func (r *FakeRuntimeService) SetFakeSandboxes(sandboxes []*FakePodSandbox) {
+	r.Lock()
+	defer r.Unlock()
+
+	r.Sandboxes = make(map[string]*FakePodSandbox)
+	for _, sandbox := range sandboxes {
+		sandboxID := sandbox.GetId()
+		r.Sandboxes[sandboxID] = sandbox
+	}
 }
 
-func (r *fakeRuntimeService) StopPodSandbox(podSandboxID string) error {
-	return fmt.Errorf("not implemented")
+func (r *FakeRuntimeService) SetFakeContainers(containers []*FakeContainer) {
+	r.Lock()
+	defer r.Unlock()
+
+	r.Containers = make(map[string]*FakeContainer)
+	for _, c := range containers {
+		containerID := c.GetName()
+		r.Containers[containerID] = c
+	}
+
+}
+
+func NewFakeRuntimeService() *FakeRuntimeService {
+	return &FakeRuntimeService{
+		Called:     make([]string, 0),
+		Containers: make(map[string]*FakeContainer),
+		Sandboxes:  make(map[string]*FakePodSandbox),
+	}
 }
 
-func (r *fakeRuntimeService) DeletePodSandbox(podSandboxID string) error {
-	return fmt.Errorf("not implemented")
+func (r *FakeRuntimeService) Version(apiVersion string) (*runtimeApi.VersionResponse, error) {
+	r.Lock()
+	defer r.Unlock()
+
+	r.Called = append(r.Called, "Version")
+
+	return &runtimeApi.VersionResponse{
+		Version:           &version,
+		RuntimeName:       &fakeRuntimeName,
+		RuntimeVersion:    &version,
+		RuntimeApiVersion: &version,
+	}, nil
+}
+
+func (r *FakeRuntimeService) CreatePodSandbox(config *runtimeApi.PodSandboxConfig) (string, error) {
+	r.Lock()
+	defer r.Unlock()
+
+	r.Called = append(r.Called, "CreatePodSandbox")
+
+	// PodSandboxID should be randomized for real container runtime, but here just use
+	// sandbox's name for easily making fake sandboxes.
+	podSandboxID := config.GetName()
+	createdAt := time.Now().Unix()
+	readyState := runtimeApi.PodSandBoxState_READY
+	r.Sandboxes[podSandboxID] = &FakePodSandbox{
+		PodSandbox: runtimeApi.PodSandbox{
+			Id:        &podSandboxID,
+			Name:      config.Name,
+			State:     &readyState,
+			CreatedAt: &createdAt,
+			Labels:    config.Labels,
+		},
+		Annotations: config.Annotations,
+	}
+
+	return podSandboxID, nil
+}
+
+func (r *FakeRuntimeService) StopPodSandbox(podSandboxID string) error {
+	r.Lock()
+	defer r.Unlock()
+
+	r.Called = append(r.Called, "StopPodSandbox")
+
+	notReadyState := runtimeApi.PodSandBoxState_NOTREADY
+	if s, ok := r.Sandboxes[podSandboxID]; ok {
+		s.State = &notReadyState
+	} else {
+		return fmt.Errorf("pod sandbox %s not found", podSandboxID)
+	}
+
+	return nil
+}
+
+func (r *FakeRuntimeService) DeletePodSandbox(podSandboxID string) error {
+	r.Lock()
+	defer r.Unlock()
+
+	r.Called = append(r.Called, "DeletePodSandbox")
+
+	if _, ok := r.Sandboxes[podSandboxID]; ok {
+		delete(r.Sandboxes, podSandboxID)
```

Comment:
nit: I don't think you need to check whether the key exists in the map or not. Just delete it.